tests/test_base_virtual_machine.cpp

```diff
@@ -137,3 +137,64 @@ TEST_F(BaseVM, get_all_ipv4_works_when_ssh_throws_executing)
     auto ipv4_count = base_vm.get_all_ipv4(key_provider);
     EXPECT_EQ(ipv4_count.size(), 0u);
 }
+
+TEST_F(BaseVM, get_all_ipv4_works)
+{
+    mpt::StubBaseVirtualMachine base_vm;
+
+    REPLACE(ssh_connect, [](auto...) { return SSH_OK; });
+    REPLACE(ssh_is_connected, [](auto...) { return true; });
+    REPLACE(ssh_channel_open_session, [](auto...) { return SSH_OK; });
+    REPLACE(ssh_userauth_publickey, [](auto...) { return SSH_OK; });
+    REPLACE(ssh_channel_is_closed, [](auto...) { return 0; });
+    REPLACE(ssh_channel_open_session, [](auto...) { return SSH_OK; });
+    REPLACE(ssh_channel_get_exit_status, [](auto...) { return SSH_OK; });
+
+    ssh_channel_callbacks callbacks{nullptr};
+    auto add_channel_cbs = [&callbacks](ssh_channel, ssh_channel_callbacks cb) {
+        callbacks = cb;
+        return SSH_OK;
+    };
+    REPLACE(ssh_add_channel_callbacks, add_channel_cbs);
+
+    int expected_status{0};
+    auto event_dopoll = [&callbacks, &expected_status](ssh_event, int timeout) {
+        if (!callbacks)
+            return SSH_ERROR;
+        callbacks->channel_exit_status_function(nullptr, nullptr, expected_status, callbacks->userdata);
+        return SSH_OK;
+    };
+    REPLACE(ssh_event_dopoll, event_dopoll);
+
+    bool ip_invoked{false};
+    std::string ip_output{"wlp4s0           UP             192.168.2.168/24 \n"
+                          "eth0             UP             192.168.2.225/24 \n"
+                          "eth1             UP             192.168.2.154/24 \n"};
+    auto remaining = ip_output.size();
+
+    auto channel_read = [&ip_output, &remaining](ssh_channel, void* dest, uint32_t count, int is_stderr, int) {
+        const auto num_to_copy = std::min(count, static_cast<uint32_t>(remaining));
+        const auto begin = ip_output.begin() + ip_output.size() - remaining;
+        std::copy_n(begin, num_to_copy, reinterpret_cast<char*>(dest));
+        remaining -= num_to_copy;
+        return num_to_copy;
+    };
+    REPLACE(ssh_channel_read_timeout, channel_read);
+
+    auto request_exec = [&ip_invoked](ssh_channel channel, const char* raw_cmd) {
+        if (raw_cmd[0] == 'i' && raw_cmd[1] == 'p')
+        {
+            ip_invoked = true;
+        }
+        return SSH_OK;
+    };
+    REPLACE(ssh_channel_request_exec, request_exec);
+
+    // Check that it doesn't throw at execution, so the test will do its job.
+    mp::SSHSession session{"host", 42};
+    EXPECT_NO_THROW(session.exec("must not throw"));
+
+    auto ipv4_count = base_vm.get_all_ipv4(key_provider);
+    EXPECT_TRUE(ip_invoked);
+    EXPECT_EQ(ipv4_count.size(), 3u);
```

Comment:
May as well check the actual IPs. We know the order, so something like `EXPECT_EQ(ipv4, std::vector<std::string>{"192...", ...})`) should be enough. 

That would cover the size and whether `ip` was invoked, so those checks could be removed and the mock for `ssh_channel_request_exec` simplified.